pkg/userdata/centos/userdata.go

```diff
@@ -0,0 +1,246 @@
+		return "", fmt.Errorf("invalid kubelet version: '%v'", err)
+	}
+	kubeletVersion := fmt.Sprintf("-%s", semverKubeletVersion.String())
+
+	dockerPackageName, err := getDockerPackageName(spec.Versions.ContainerRuntime.Version)
+	if err != nil {
+		return "", fmt.Errorf("error getting Docker package name: '%v'", err)
+	}
+
+	cpConfig, cpName, err := ccProvider.GetCloudConfig(spec)
+	if err != nil {
+		return "", fmt.Errorf("failed to get cloud config: %v", err)
+	}
+
+	pconfig, err := providerconfig.GetConfig(spec.ProviderConfig)
+	if err != nil {
+		return "", fmt.Errorf("failed to get provider config: %v", err)
+	}
+
+	osConfig, err := getConfig(pconfig.OperatingSystemSpec)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse OperatingSystemSpec: '%v'", err)
+	}
+
+	data := struct {
+		MachineSpec       machinesv1alpha1.MachineSpec
+		ProviderConfig    *providerconfig.Config
+		OSConfig          *Config
+		Kubeconfig        string
+		CloudProvider     string
+		CloudConfig       string
+		KubeletVersion    string
+		DockerPackageName string
+		ClusterDNSIPs     []net.IP
+		KubernetesCACert  string
+	}{
+		MachineSpec:       spec,
+		ProviderConfig:    pconfig,
+		OSConfig:          osConfig,
+		Kubeconfig:        kubeconfig,
+		CloudProvider:     cpName,
+		CloudConfig:       cpConfig,
+		KubeletVersion:    kubeletVersion,
+		DockerPackageName: dockerPackageName,
+		ClusterDNSIPs:     clusterDNSIPs,
+		KubernetesCACert:  kubernetesCACert,
+	}
+	b := &bytes.Buffer{}
+	err = tmpl.Execute(b, data)
+	if err != nil {
+		return "", fmt.Errorf("failed to execute user-data template: %v", err)
+	}
+	return b.String(), nil
+}
+
+const ctTemplate = `#cloud-config
+hostname: {{ .MachineSpec.Name }}
+
+{{- if .OSConfig.DistUpgradeOnBoot }}
+package_upgrade: true
+package_reboot_if_required: true
+{{- end }}
+
+{{ if ne (len .ProviderConfig.SSHPublicKeys) 0 }}
+ssh_authorized_keys:
+{{- range .ProviderConfig.SSHPublicKeys }}
+  - "{{ . }}"
+{{- end }}
+{{- end }}
+
+write_files:
+- path: "/etc/yum.repos.d/kubernetes.repo"
+  content: |
+    [kubernetes]
+    name=Kubernetes
+    baseurl=https://packages.cloud.google.com/yum/repos/kubernetes-el7-$basearch
+    enabled=1
+    gpgcheck=1
+    repo_gpgcheck=1
+    gpgkey=https://packages.cloud.google.com/yum/doc/yum-key.gpg https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg
+
+- path: /etc/sysconfig/selinux
+  content: |
+    # This file controls the state of SELinux on the system.
+    # SELINUX= can take one of these three values:
+    #     enforcing - SELinux security policy is enforced.
+    #     permissive - SELinux prints warnings instead of enforcing.
+    #     disabled - No SELinux policy is loaded.
+    SELINUX=permissive
+    # SELINUXTYPE= can take one of three two values:
+    #     targeted - Targeted processes are protected,
+    #     minimum - Modification of targeted policy. Only selected processes are protected.
+    #     mls - Multi Level Security protection.
```

Comment:
I assume those are comments?
Are those necessary?

Reply:
Yes those are comments. They are not necessary for functionality but they are part how the file looks by default